actix-server/src/ssl/nativetls.rs

```diff
@@ -68,130 +80,46 @@ pub struct NativeTlsAcceptorService<T, P> {
+where
+    T: AsyncRead + AsyncWrite + Unpin + 'static,
+    P: 'static,
+{
+    type Request = Io<T, P>;
+    type Response = Io<TlsStream<T>, P>;
     type Error = Error;
+    type Future = LocalBoxFuture<'static, Result<Io<TlsStream<T>, P>, Error>>;
 
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        match self.inner.take().expect("cannot poll MidHandshake twice") {
-            Ok(stream) => Ok(Async::Ready(Io::from_parts(
-                TlsStream { inner: stream },
-                self.params.take().unwrap(),
-                Protocol::Unknown,
-            ))),
-            Err(HandshakeError::Failure(e)) => Err(e),
-            Err(HandshakeError::WouldBlock(s)) => match s.handshake() {
-                Ok(stream) => Ok(Async::Ready(Io::from_parts(
-                    TlsStream { inner: stream },
-                    self.params.take().unwrap(),
-                    Protocol::Unknown,
-                ))),
-                Err(HandshakeError::Failure(e)) => Err(e),
-                Err(HandshakeError::WouldBlock(s)) => {
-                    self.inner = Some(Err(HandshakeError::WouldBlock(s)));
-                    Ok(Async::NotReady)
-                }
-            },
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        if self.conns.available(cx) {
+            Poll::Ready(Ok(()))
+        } else {
+            Poll::Pending
         }
     }
-}
-
-impl<S> TlsStream<S> {
-    /// Get access to the internal `native_tls::TlsStream` stream which also
-    /// transitively allows access to `S`.
-    pub fn get_ref(&self) -> &native_tls::TlsStream<S> {
-        &self.inner
-    }
-
-    /// Get mutable access to the internal `native_tls::TlsStream` stream which
-    /// also transitively allows mutable access to `S`.
-    pub fn get_mut(&mut self) -> &mut native_tls::TlsStream<S> {
-        &mut self.inner
-    }
-}
-
-impl<S: io::Read + io::Write> io::Read for TlsStream<S> {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        self.inner.read(buf)
-    }
-}
-
-impl<S: io::Read + io::Write> io::Write for TlsStream<S> {
-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        self.inner.write(buf)
-    }
 
-    fn flush(&mut self) -> io::Result<()> {
-        self.inner.flush()
-    }
-}
-
-impl<S: AsyncRead + AsyncWrite> AsyncRead for TlsStream<S> {}
-
-impl<S: AsyncRead + AsyncWrite> AsyncWrite for TlsStream<S> {
-    fn shutdown(&mut self) -> Poll<(), io::Error> {
-        match self.inner.shutdown() {
-            Ok(_) => (),
-            Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => (),
-            Err(e) => return Err(e),
-        }
-        self.inner.get_mut().shutdown()
+    fn call(&mut self, req: Self::Request) -> Self::Future {
+        let guard = self.conns.get();
+        let this = self.clone();
+        let (io, params, proto) = req.into_parts();
+        async move { this.acceptor.accept(io).await }
```

Comment:
@fafhrd91 here is the place I don't really like, because it imposes unnecessary `'static` bounds on type parameters and requires additional cloning. We're made to do so, however, because of this method signature:
```rust
fn call(&mut self, req: Self::Request) -> Self::Future
```

In fact, if Rust would have supported `async fn` in traits, it would be
```rust
async fn call(&mut self, req: Self::Request) -> Result<Self::Response, Self::Error>
```
and in such a way would require no `'static`/`.clone()`ing at all.

Seems being same, but the point is in lifetimes elision rules. `async fn` output captures lifetimes of all its parameters lifetimes into the output, while our function does not capture any which results in "outlive" error.

Would it be acceptable to bikeshed lifetimes of `Service::call()` method to make it more usable along with `async fn`s?

Reply:
we can bikeshed, but i dont have any idea about async call.

Reply:
i dont think you can remove extra clone. borrow checker is still there, so you would need to do clone or framework code would do that for you.

Reply:
Yeah, I've realized that `.clone()` happens because of `LocalBoxFuture` having lifetime parameter. And as nor Rust supports `'self` lifetimes, neither we have lifetime parameter in type signature, we have to use only `'static` there, so that's a definitely requires clone.